corehq/apps/app_manager/tests/test_views.py

```diff
@@ -223,32 +224,44 @@ def test_default_new_app(self, mock):
         self.addCleanup(lambda: Application.get_db().delete_doc(app_id))
 
     def test_get_apps_modules(self, mock):
-        self.app.add_module(Module.new_module("Module0", "en"))
-        self.app.save()
-
-        other_app = Application.new_app(self.project.name, "OtherApp")
-        other_app.add_module(Module.new_module("Module0", "en"))
-        other_app.save()
-        self.addCleanup(lambda: Application.get_db().delete_doc(other_app.id))
-
-        linked_app = create_linked_app(self.project.name, self.app.id, self.project.name, 'LinkedApp')
-        self.addCleanup(lambda: Application.get_db().delete_doc(linked_app.id))
-
-        deleted_app = Application.new_app(self.project.name, "DeletedApp")
-        deleted_app.add_module(Module.new_module("Module0", "en"))
-        deleted_app.save()
-        deleted_app.delete_app()
-        deleted_app.save()
-        self.addCleanup(lambda: Application.get_db().delete_doc(deleted_app.id))
-
-        apps_modules = get_apps_modules(self.project.name)
-
-        self.assertEqual(len(apps_modules), 2, 'get_apps_modules should only return normal Applications')
-        self.assertTrue(
-            all(len(app['modules']) == 1 for app in apps_modules),
-            'Each app should only have one module'
-        )
-        self.assertEqual(
-            apps_modules[0]['modules'][0]['name'], 'Module0',
-            'Module name should be translated'
-        )
+        with apps_modules_setup(self):
+            apps_modules = get_apps_modules(self.project.name)
+
+            self.assertEqual(len(apps_modules), 2, 'get_apps_modules should only return normal Applications')
+            self.assertTrue(
+                all(len(app['modules']) == 1 for app in apps_modules),
+                'Each app should only have one module'
+            )
+            self.assertEqual(
+                apps_modules[0]['modules'][0]['name'], 'Module0',
+                'Module name should be translated'
+            )
+
+    def test_get_apps_modules_doc_types(self, mock):
+        with apps_modules_setup(self):
+            apps_modules = get_apps_modules(
+                self.project.name, app_doc_types=('Application', 'LinkedApplication')
+            )
+            names = sorted([a['name'] for a in apps_modules])
+            self.assertEqual(names, ['LinkedApp', 'OtherApp', 'TestApp'])
+
+
+@contextmanager
+def apps_modules_setup(test_case):
```

Comment:
I chose to use a context manager for setup and teardown for these two tests, because DRY. Another option would be to subclass the test case, but that would run the superclass's tests twice.

Reply:
Interesting.. what do you mean by `run the superclass's tests twice.`?

Reply:
Might be easiest with an example:
```python
from unittest import TestCase


class FooTest(TestCase):

    def setUp(self):
        self.foo = 2

    def test_foo_two(self):
        self.assertEqual(self.foo, 2)


class BarTest(FooTest):

    def setUp(self):
        super(BarTest, self).setUp()
        self.bar = 1

    def test_bar_one(self):
        self.assertEqual(self.bar, 1)
```
So the `BarTest` class extends the `FooTest` test case, and we are doing this, say, to add additional setup for testing `bar`. When we run this we can see that, because `BarTest` inherits all the tests in `FooTest`, it runs `test_foo_two` twice, once for `FooTest`, and again for `BarTest`:
```bash
    $ python -m unittest -v test_example
    test_bar_one (test_example.BarTest) ... ok
    test_foo_two (test_example.BarTest) ... ok
    test_foo_two (test_example.FooTest) ... ok

    ----------------------------------------------------------------------
    Ran 3 tests in 0.000s

    OK
```
So to avoid running the same tests more than once, but still setting up and tearing down extra fixtures, context managers are great. (In fact, there is a Python [test framework](https://pythonhosted.org/Attest/) that uses context managers instead of setUp and tearDown methods.)

Reply:
Interesting.. Thanks for sharing this @kaapstorm . I am not sure if we follow that pattern at places in our tests and that would mean its running all those tests twice/multiple times on travis?
Probably one way to avoid this would be to have the base class do only setup and teardown and add no tests unless they are to just test the setup itself which would be a valid test for all subclasses.
(non blocking for this PR)